kernel/vfs/src/node/directory.rs

```diff
@@ -0,0 +1,147 @@
+// TODO: We would rather these methods take &Arc<Self>, but that isn't currently
+// possible.
+// https://internals.rust-lang.org/t/rc-arc-borrowed-an-object-safe-version-of-rc-t-arc-t/8896/9
+
+use crate::{File, Node, NodeKind, Path};
+use alloc::{sync::Arc, vec::Vec};
+
+/// A directory.
+///
+/// Consumers of this trait should use the methods implemented for the `dyn
+/// Directory` type.
+pub trait Directory: Node {
+    /// Retrieves the entry with the specified name.
+    ///
+    /// `name` must not contain slashes.
+    fn get_entry(&self, name: &str) -> Option<Arc<dyn Node>>;
+
+    /// Retrieves the file entry with the specified name.
+    ///
+    /// `name` must not contain slashes.
+    fn get_file_entry(&self, name: &str) -> Option<Arc<dyn File>> {
+        match self.get_entry(name)?.as_specific() {
+            NodeKind::File(file) => Some(file),
+            _ => None,
+        }
+    }
+
+    /// Retrieves the directory entry with the specified name.
+    ///
+    /// `name` must not contain slashes.
+    fn get_directory_entry(&self, name: &str) -> Option<Arc<dyn Directory>> {
+        match self.get_entry(name)?.as_specific() {
+            NodeKind::Directory(dir) => Some(dir),
+            _ => None,
+        }
+    }
+
+    /// Inserts a file with the specified name, returning it.
+    ///
+    /// `name` must not contain slashes.
+    fn insert_file_entry(self: Arc<Self>, name: &str) -> Result<Arc<dyn File>, &'static str>;
+
+    /// Inserts a directory with the specified name, returning it.
+    ///
+    /// `name` must not contain slashes.
+    fn insert_directory_entry(
+        self: Arc<Self>,
+        name: &str,
+    ) -> Result<Arc<dyn Directory>, &'static str>;
```

Comment:
Having a common `insert_entry` (method taking a NodeKind) could be good here too, no?

Reply:
The API intentionally doesn't let you insert an existing `dyn Node` (or `dyn File`/`dyn Directory`/`NodeKind`) so that different file system nodes aren't mixed.

Reply:
> On another note I think create is a better name than insert.
I'd say "insert" refers to slot insertion (one may think it implies there's a growable array underneath) whereas "create" is more general/natural.